scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+                                 self.samples_per_frame, u.s)
+        return 1.0 / sample_time
+
+    @sample_rate.setter
+    def sample_rate(self, val):
+        if hasattr(val, 'unit'):
+            val = val.to(u.Hz).value
+        self.header['TBIN'] = 1.0 / val
+
+    @property
+    def start_time(self):
+        # NOTE should we get the start time for each raw, in case the time gaps
+        # in between the rows
+        file_start = self.header_hdu.start_time
+        if "OFFS_SUB" in self.columns.names:
+            subint_times = u.Quantity(self.data['OFFS_SUB'], u.s, copy=False)
+            sample_time = 1.0 / self.sample_rate
+            start_time = (file_start + subint_times[0] -
+                          self.samples_per_frame / 2 * sample_time)
+        else:
+            start_time = file_start
+        return start_time
+
+    @start_time.setter
+    def start_time(self, time):
+        # NOTE this sets the start time of the HDU, not the file start time.
+        dt = (time - self.header_hdu.start_time).to(u.s)
+        center_off = dt + 1.0 / self.sample_rate * self.samples_per_frame / 2
+        if "OFFS_SUB" in self.columns.names:
+            first_off = self.data['OFFS_SUB'][0]
+            self.data['OFFS_SUB'] = (self.data['OFFS_SUB'] - first_off +
+                                     center_off)
+        else:
+            self._req_columns['OFFS_SUB']['value'] = center_off
+
+    @property
+    def nrow(self):
+        return int(self.header['NAXIS2'])
+
+    @property
+    def nchan(self):
+        return int(self.header['NCHAN'])
+
+    @property
+    def npol(self):
+        return int(self.header['NPOL'])
+
+    @property
+    def nbin(self):
+        return int(self.header['NBIN'])
+
+    @property
+    def samples_per_frame(self):
+        try:
+            return int(self.header['NSBLK'])
+        except:
+            return int(np.prod(self.data_shape)/(self.nrow * self.nchan *
+                                                 self.npol * self.nbin))
+
+    @samples_per_frame.setter
+    def samples_per_frame(self, val):
+        self.header['NSBLK'] = val
+
+    @property
+    def shape(self):
+        raw_shape = self.raw_shape
+        new_shape = namedtuple('shape', ['nsample','nbin', 'nchan', 'npol'])
+        result = new_shape(raw_shape.nrow * raw_shape.samples_per_frame,
+                           raw_shape.nbin, raw_shape.nchan, raw_shape.npol)
+        return result
+
+    @property
+    def raw_shape(self):
+        r_shape = namedtuple('shape', ['nrow', 'samples_per_frame','nbin',
+                                       'nchan','npol'])
+        result = r_shape(self.nrow, self.samples_per_frame, self.nbin,
+                         self.nchan, self.npol)
+        return result
+
+    @property
+    def data_shape(self):
+        # Data are save in the fortran order. Reversed from the header label.
+        d_shape_raw = self.data['DATA'].shape
+        if self.mode == "SEARCH":
```

Comment:
This if/else suggests we should have subclasses for SEARCH and other types.

Reply:
I agree, I can image we will use SUBINT for some other type of data.